### PR TITLE
(WIP) Allow requires_grad=False for forward AD gradcheck

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #72271
* #72166
* #72165

Notes in case this gets picked up later:
- adds differentiable inputs/outputs indices parameter to OpInfos and gradcheck
  - gradcheck only accepts positional arguments, but sample inputs can have kwargs, so we need to transform those
- we have two wrapper functions, `fn_fw` and `fn_rev`, the only difference being that `fn_rev` filters out inputs/outputs that don't require grad
- check non differentiable inputs (only applies to reverse-mode) needs to use `fn_fw` because it needs to check that the outputs that don't have requires_grad=True have numerical Jacobian of zero
- we cannot remove the checks for get_numerical_jacobian because it is considered external API, unfortunately. They are redundant, but don't cause any harm.
- we cannot remove iter_tensors because it is still used in common_nn
- add some more asserts to make sure there is no silent size mismatch, including zip -> _safe_zip 
- example op, batch norm. There doesn't seem like any others where differentiability could not be inferred from dtype, or no error was produced

Differential Revision: [D33991621](https://our.internmc.facebook.com/intern/diff/D33991621)

### CODE REVIEW

**Comment by soulitzer** [comment]
@soulitzer has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D33991621).

**Comment by soulitzer** [comment]
Closing for now, because there are cheaper options. Ideally, we'd want to move in the direction of not using the requires_grad trick to mark inputs/outputs of a op as differentiable or not.
